### PR TITLE
Refactor of ETH (and RSK) Plugin

### PR DESCRIPTION
**Pull request by kylanhurt**

### CODE REVIEW

File: src/ethereum/ethNetwork.js [hunk 1]
```diff
@@ -751,14 +795,21 @@ export class EthereumNetwork {
...
-        funcs2 = async () => {
-          const result = await this.fetchPostInfura('eth_getTransactionCount', [
-            params[0],
-            'latest'
-          ])
-          return { server: 'infura', result }
+        if (infuraServers.length > 0) {
+          funcs2 = async () => {
+            const adjustedParams =
+              this.currencyInfo.currencyCode === 'RBTC'
```
Comment (kylanhurt) [inline src/ethereum/ethNetwork.js]: Wondering what you guys think of the reference to 'RBTC'. For some reason their Infura nodes expect the `param` array to be nested within an array (at least that's what works with the node that we are using).

Comment (paullinator) [reply src/ethereum/ethNetwork.js]: Yuck. Put a flag in the currenyInfo

File: src/ethereum/ethNetwork.js [hunk 2]
```diff
@@ -890,11 +947,22 @@ export class EthereumNetwork {
...
-      this.checkBlockHeightAmberdata,
-      this.checkBlockHeightBlockchair
-    ]).catch(err => {
+    const {
+      etherscanApiServers,
+      amberdataApiServers,
+      blockchairApiServers
+    } = this.currencyInfo.defaultSettings.otherSettings
+    const waterfallFuncs = []
+    if (etherscanApiServers.length > 0) {
```
Comment (paullinator) [inline src/ethereum/ethNetwork.js]: The IF statements should not be necessary here

File: src/ethereum/ethNetwork.js [hunk 3]
```diff
@@ -931,10 +999,19 @@ export class EthereumNetwork {
...
-    return asyncWaterfall([
-      this.checkNonceEthscan,
-      this.checkNonceAmberdata
-    ]).catch(err => {
+    const {
+      etherscanApiServers,
+      amberdataApiServers
+    } = this.currencyInfo.defaultSettings.otherSettings
+    const waterfallFuncs = []
+    if (etherscanApiServers.length > 0) {
```
Comment (paullinator) [inline src/ethereum/ethNetwork.js]: The IF statements should not be necessary here

File: src/ethereum/ethNetwork.js [hunk 4]
```diff
@@ -1315,21 +1395,26 @@ export class EthereumNetwork {
...
   }
 
   async checkTokenBalBlockchair(): Promise<EthereumNetworkUpdate> {
     const address = this.ethEngine.walletLocalData.publicKey
+    const {
+      blockchairUrlTokenString
+    } = this.currencyInfo.defaultSettings.otherSettings
     const jsonObj = await this.fetchGetBlockchair(
-      `/ethereum/dashboards/address/${address}?erc_20=true`,
+      `/${this.currencyInfo.pluginName}/dashboards/address/${address}?${blockchairUrlTokenString}=true`,
```
Comment (paullinator) [inline src/ethereum/ethNetwork.js]: blockchairUrlTokenString can be assumed to be `erc_20` all the time

File: src/ethereum/ethNetwork.js [hunk 5]
```diff
@@ -1350,10 +1435,15 @@ export class EthereumNetwork {
...
-    return asyncWaterfall([
-      async () => this.checkTokenBalEthscan(tk),
-      this.checkTokenBalBlockchair
-    ]).catch(err => {
+    const {
+      blockchairApiServers
+    } = this.currencyInfo.defaultSettings.otherSettings
+    const waterfallFuncs = []
+    waterfallFuncs.push(async () => this.checkTokenBalEthscan(tk))
+    if (blockchairApiServers.length > 0) {
```
Comment (paullinator) [inline src/ethereum/ethNetwork.js]: The IF statements should not be necessary here

File: src/ethereum/ethPlugin.js [hunk 6]
```diff
@@ -63,195 +47,583 @@ const defaultNetworkFees = {
...
-    opts: EdgeCurrencyEngineOptions
-  ): Promise<EdgeCurrencyEngine> {
-    const tools = await makeCurrencyTools()
-    const currencyEngine = new EthereumEngine(
-      tools,
-      walletInfo,
-      initOptions,
-      opts
-    )
+export const currencyInfo: EdgeCurrencyInfo = {
```
Comment (paullinator) [inline src/ethereum/ethPlugin.js]: Keep this in an ethInfo.js file

File: src/ethereum/rskPlugin.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  EdgeCorePluginOptions,
+  EdgeCurrencyInfo
+} from 'edge-core-js/types'
+
+import { makeEthereumBasedPluginInner } from './ethBasedPlugin.js'
+import type { EthereumSettings } from './ethTypes.js'
+
+export const imageServerUrl = 'https://developer.airbitz.co/content'
+
+const defaultNetworkFees = {
```
Comment (paullinator) [inline src/ethereum/rskPlugin.js]: Keep this in an rskInfo.js file

**Comment by alepc253** [comment]
I tested it locally and seems to be working OK

File: src/ethereum/ethEngine.js [hunk 8]
```diff
@@ -278,10 +289,15 @@ export class EthereumEngine extends CurrencyEngine {
   async startEngine() {
     this.engineOn = true
     this.addToLoop('checkUpdateNetworkFees', NETWORKFEES_POLL_MILLISECONDS)
-    this.addToLoop(
-      'checkUnconfirmedTransactionsInnerLoop',
-      UNCONFIRMED_TRANSACTION_POLL_MILLISECONDS
-    )
+    if (
+      this.currencyInfo.defaultSettings.otherSettings
```
[the remark this answers is not in the document]
Comment (kylanhurt) [reply src/ethereum/ethEngine.js]: @paullinator I don't see anywhere in the code to checks that there actually is a URL for this call before deciding whether to do the `fetch` call or not. Here is the code in `ethEngine`:
```
    const url = `${this.currencyInfo.defaultSettings.otherSettings.superethServers[0]}/v1/${lowerCaseCurrencyCode}/main/txs/${address}`
    let jsonObj = null
    try {
      jsonObj = await this.ethNetwork.fetchGet(url)
```

Can I use the length of the `superethServers` array as the flag for whether or not to fire off the fetch routine?

File: src/ethereum/ethNetwork.js [hunk 9]
```diff
@@ -671,142 +708,169 @@ export class EthereumNetwork {
...
-        promises.push(
-          broadcastWrapper(this.broadcastInfura(params[0]), 'infura')
-        )
-        promises.push(
-          broadcastWrapper(this.broadcastEtherscan(params[0]), 'etherscan')
-        )
-        promises.push(
-          broadcastWrapper(this.broadcastBlockCypher(params[0]), 'blockcypher')
-        )
+        if (infuraServers.length > 0) {
```
[the remark this answers is not in the document]
Comment (paullinator) [reply src/ethereum/ethNetwork.js:741]: Look at how eth_getBalance works

File: src/ethereum/ethNetwork.js [hunk 10]
```diff
@@ -163,6 +161,7 @@ export class EthereumNetwork {
       if (tx.gasPrice) {
         nativeNetworkFee = bns.mul(tx.gasPrice, tx.gasUsed)
       } else {
+        // implemented for ETH, not RSK
```
[the remark this answers is not in the document]
Comment (kylanhurt) [reply src/ethereum/ethNetwork.js:165]: It just means that the `else` clause existed in the Ethereum version of the function, but not the RSK version, before the refactor. So if there was no `tx.contractAddress` and no `tx.gasPrice` then `nativeNetworkFee` was not being assigned to `'0'` for RSK before.

**Pull request merged**